File: apps/ai-api/routes/llm.py
"""
LLM endpoints - chat and model comparison
"""
import json
import time
import uuid

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from models.llm import ChatRequest, CompareRequest
from services.llm import format_chat_prompt, generate_llm_stream
from state import llm_engines, model_info

router = APIRouter(prefix="/api", tags=["LLM"])


@router.get(
    "/tags",
    summary="List models",
    description="Returns a list of available LLM models",
)
async def list_models():
    """List available LLM models"""
    models = []
    for model_id, info in model_info.items():
        models.append({
            "name": info["name"],
            "size": info["size"],
            "modified_at": "",
        })
    return {"models": models}


@router.post(
    "/chat",
    summary="Chat completion",
    description="Generate a chat completion using the specified model",
)
async def chat(request: ChatRequest):
    """Chat with a model"""
    from vllm import SamplingParams

    # Find engine by model name
    engine = None
    model_id = None

    for mid, eng in llm_engines.items():
        if request.model in mid or request.model == model_info.get(mid, {}).get("name"):
            engine = eng
            model_id = mid
            break

    if not engine:
        raise HTTPException(status_code=404, detail=f"Model {request.model} not found")

    prompt = format_chat_prompt(request.messages, model_id)

    sampling_params = SamplingParams(
        temperature=request.temperature,
        top_p=request.top_p,
        top_k=request.top_k,
        max_tokens=request.max_tokens,
    )

    if request.stream:
        return StreamingResponse(
            generate_llm_stream(engine, prompt, sampling_params, request.model),
            media_type="text/event-stream",
        )

    # Non-streaming response
    request_id = str(uuid.uuid4())
    start_time = time.time()

    results = []
    async for output in engine.generate(prompt, sampling_params, request_id):
        results.append(output)

    final_output = results[-1] if results else None

    if not final_output or not final_output.outputs:
        raise HTTPException(status_code=500, detail="Generation failed")

    generated_text = final_output.outputs[0].text
    total_duration = int((time.time() - start_time) * 1e9)

    return {
        "model": request.model,
        "message": {
            "role": "assistant",
            "content": generated_text,
        },
        "done": True,
        "total_duration": total_duration,
        "eval_count": len(generated_text.split()),
    }


@router.post(
    "/compare",
    summary="Compare models",
    description="Compare responses from multiple models for the same prompt",
)
async def compare_models(request: CompareRequest):
    """Compare multiple models (streaming)"""
    from vllm import SamplingParams

    async def generate_comparison():
        start_time = time.time()

        prompt = format_chat_prompt(request.messages, "")

        sampling_params = SamplingParams(
            temperature=request.temperature,
            top_p=request.top_p,
            top_k=request.top_k,
            max_tokens=request.max_tokens,
        )

        for model_name in request.models:
            engine = None

            for mid, eng in llm_engines.items():
                if model_name in mid or model_name == model_info.get(mid, {}).get("name"):
                    engine = eng
                    break

            if not engine:
                data = json.dumps({
                    "model": model_name,
                    "error": "Model not found",
                    "done": True
                })
                yield f"event: model_error\ndata: {data}\n\n"
                continue

            model_start_time = time.time()
            request_id = str(uuid.uuid4())
            full_content = ""

            async for request_output in engine.generate(prompt, sampling_params, request_id):
                if request_output.outputs:
                    output = request_output.outputs[0]
                    new_text = output.text[len(full_content):]
                    full_content = output.text

                    if new_text:
                        data = json.dumps({
                            "model": model_name,
                            "content": new_text,
                            "done": False
                        })
                        yield f"event: chunk\ndata: {data}\n\n"

            duration = int((time.time() - model_start_time) * 1000)
            data = json.dumps({
                "model": model_name,
                "fullContent": full_content,
                "duration": duration
            })
            yield f"event: model_done\ndata: {data}\n\n"

        total_duration = int((time.time() - start_time) * 1000)
        data = json.dumps({"totalDuration": total_duration})
        yield f"event: all_done\ndata: {data}\n\n"

    return StreamingResponse(
        generate_comparison(),
        media_type="text/event-stream",
    )

File: apps/server/src/routes/llm.ts
import { zValidator } from "@hono/zod-validator";
import { Hono } from "hono";
import { streamSSE } from "hono/streaming";
import { z } from "zod";

const AI_URL = process.env.AI_URL || "http://localhost:8000";

const llm = new Hono();

// Schemas
const chatSchema = z.object({
  model: z.string(),
  messages: z.array(
    z.object({
      role: z.enum(["system", "user", "assistant"]),
      content: z.string(),
    })
  ),
  stream: z.boolean().optional().default(true),
  options: z
    .object({
      temperature: z.number().optional(),
      top_p: z.number().optional(),
      top_k: z.number().optional(),
    })
    .optional(),
});

const compareSchema = z.object({
  models: z.array(z.string()).min(1).max(5),
  messages: z.array(
    z.object({
      role: z.enum(["system", "user", "assistant"]),
      content: z.string(),
    })
  ),
  options: z
    .object({
      temperature: z.number().optional(),
      top_p: z.number().optional(),
      top_k: z.number().optional(),
    })
    .optional(),
});

type ChatMessage = {
  role: "system" | "user" | "assistant";
  content: string;
};

type OllamaResponse = {
  model: string;
  message: ChatMessage;
  done: boolean;
  total_duration?: number;
  eval_count?: number;
};

// Get available models
llm.get("/models", async (c) => {
  const response = await fetch(`${AI_URL}/api/tags`);

  if (!response.ok) {
    return c.json({ error: "Failed to fetch models from LLM service" }, 500);
  }

  const data = (await response.json()) as {
    models: Array<{ name: string; size: number; modified_at: string }>;
  };

  return c.json({
    models: data.models.map((m) => ({
      name: m.name,
      size: m.size,
      modified_at: m.modified_at,
    })),
  });
});

// Chat with single model (streaming)
llm.post("/chat", zValidator("json", chatSchema), async (c) => {
  const body = c.req.valid("json");

  if (body.stream) {
    return streamSSE(c, async (stream) => {
      const response = await fetch(`${AI_URL}/api/chat`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          model: body.model,
          messages: body.messages,
          stream: true,
          options: body.options,
        }),
      });

      if (!(response.ok && response.body)) {
        await stream.writeSSE({
          data: JSON.stringify({ error: "Failed to connect to LLM service" }),
          event: "error",
        });
        return;
      }

      const reader = response.body.getReader();
      const decoder = new TextDecoder();

      let buffer = "";
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() || "";

        for (const line of lines) {
          if (line.trim()) {
            const data = JSON.parse(line) as OllamaResponse;
            await stream.writeSSE({
              data: JSON.stringify({
                content: data.message?.content || "",
                done: data.done,
                model: data.model,
              }),
              event: "message",
            });
          }
        }
      }

      await stream.writeSSE({ data: "[DONE]", event: "done" });
    });
  }

  // Non-streaming response
  const response = await fetch(`${AI_URL}/api/chat`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      model: body.model,
      messages: body.messages,
      stream: false,
      options: body.options,
    }),
  });

  if (!response.ok) {
    return c.json({ error: "Failed to get response from LLM service" }, 500);
  }

  const data = (await response.json()) as OllamaResponse;

  return c.json({
    model: data.model,
    message: data.message,
    total_duration: data.total_duration,
    eval_count: data.eval_count,
  });
});

// Compare multiple models (streaming)
llm.post("/compare", zValidator("json", compareSchema), async (c) => {
  const body = c.req.valid("json");

  return streamSSE(c, async (stream) => {
    const startTime = Date.now();

    // Start all model requests in parallel
    const modelPromises = body.models.map(async (model) => {
      const modelStartTime = Date.now();

      const response = await fetch(`${AI_URL}/api/chat`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          model,
          messages: body.messages,
          stream: true,
          options: body.options,
        }),
      });

      if (!(response.ok && response.body)) {
        await stream.writeSSE({
          data: JSON.stringify({
            model,
            error: "Failed to connect",
            done: true,
          }),
          event: "model_error",
        });
        return;
      }

      const reader = response.body.getReader();
      const decoder = new TextDecoder();

      let buffer = "";
      let fullContent = "";

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() || "";

        for (const line of lines) {
          if (line.trim()) {
            const data = JSON.parse(line) as OllamaResponse;
            const content = data.message?.content || "";
            fullContent += content;

            await stream.writeSSE({
              data: JSON.stringify({
                model,
                content,
                done: data.done,
              }),
              event: "chunk",
            });

            if (data.done) {
              const duration = Date.now() - modelStartTime;
              await stream.writeSSE({
                data: JSON.stringify({
                  model,
                  fullContent,
                  duration,
                  eval_count: data.eval_count,
                }),
                event: "model_done",
              });
            }
          }
        }
      }
    });

    await Promise.all(modelPromises);

    const totalDuration = Date.now() - startTime;
    await stream.writeSSE({
      data: JSON.stringify({ totalDuration }),
      event: "all_done",
    });
  });
});

export default llm;
